src/select.test.ts

```diff
@@ -0,0 +1,222 @@
+import { q } from "./index";
+import { z } from "zod";
+import { describe, expect, expectTypeOf, it } from "vitest";
+
+describe("q.select()", () => {
+  it("creates query from {condition: selection} composition", () => {
+    const { query, schema } = q.__experimental_select({
+      "foo > 2": {
+        bar: z.boolean(),
+      },
+      default: {
+        baz: z.string(),
+      },
+    });
+
+    expect(query).toBe("select(foo > 2 => { bar }, { baz })");
+    expectTypeOf({} as z.infer<typeof schema>).toEqualTypeOf<
+      { bar: boolean } | { baz: string }
+    >();
+  });
+
+  it("creates query from {condition: q()} composition", () => {
+    const { query, schema } = q.__experimental_select({
+      "foo > 2": q("bar").grab({
+        bar: z.boolean(),
+      }),
+      default: {
+        baz: z.string(),
+      },
+    });
+
+    expect(query).toBe("select(foo > 2 => bar{bar}, { baz })");
+    expectTypeOf({} as z.infer<typeof schema>).toEqualTypeOf<
+      { bar: boolean } | { baz: string }
+    >();
+  });
+
+  it("creates query from {condition: [name, schema]} composition", () => {
+    const { query, schema } = q.__experimental_select({
+      "foo > 2": ["bar", q.boolean()],
+      default: {
+        baz: z.string(),
+      },
+    });
+
+    expect(query).toBe("select(foo > 2 => bar, { baz })");
+    expectTypeOf({} as z.infer<typeof schema>).toEqualTypeOf<
+      boolean | { baz: string }
+    >();
+  });
+
+  it("makes schema nullable when default condition omitted", () => {
+    const { query, schema } = q.__experimental_select({
+      "foo > 2": {
+        bar: z.boolean(),
+      },
+    });
+
+    expect(query).toBe("select(foo > 2 => { bar })");
+    expectTypeOf({} as z.infer<typeof schema>).toEqualTypeOf<{
+      bar: boolean;
+    } | null>();
+  });
+
+  it("handles nested selects properly", () => {
+    const nestedSelect = q.__experimental_select({
+      'bar == "thing"': {
+        b: q.string(),
+      },
+      default: {
+        c: z.string(),
+      },
+    });
+
+    const { query, schema } = q.__experimental_select({
+      "foo > 2": nestedSelect,
+      default: {
+        baz: z.string(),
+      },
+    });
+
+    expect(query).toBe(
+      'select(foo > 2 => select(bar == "thing" => { b }, { c }), { baz })'
+    );
+    expectTypeOf({} as z.infer<typeof schema>).toEqualTypeOf<
+      { b: string } | { c: string } | { baz: string }
+    >();
+  });
+});
+
+describe("EntityQuery.select()", () => {
+  describe("with standalone select input", () => {
+    it("updates query and handles nested object types", () => {
+      const standaloneSelect = q.__experimental_select({
+        "foo > 2": {
+          bar: z.boolean(),
+        },
+        default: {
+          baz: z.string(),
+        },
+      });
+      const entityQuery = q("foo").__experimental_select(standaloneSelect);
```

Comment:
~~So when you call the select method with an instance of the select method's return type, it does the spread behavior? If that's the case, it seems like a pretty obscure way to trigger that behavior (= hard to read the code, hard to write the code).~~
~~Why not have a separate method called `selectAndSpread`/`spreadSelect` or something like that, so you could do the following:~~
```js
q('foo').spreadSelect({
  "foo > 2": {
    bar: z.boolean(),
  },
  default: {
    baz: z.string(),
  },
})
```

Reply:
I think I see now, you were just demonstrating that the `.select()`'s return type works just fine when you plug them back into a bigger query via another `.select()`. This API is allowed for reusability (yeah, I read your PR notes, but didn't quite grok it at the time). Whether to spread or not is sensitive to the context in which the select is called.

Reply:
yeaup. exactly. Ideally groq would allow something like this:
```
*[]select(...)
```
but it only allows `select` inside a block context, so we get something similar by doing this instead:
```
*[]{
  ...select(...)
}  
```


In the current api prior to this PR, you would accomplish this through the second argument of the `grab` call ie:
```ts
const baseSelection = {
  _type: q.literal('someEntity')
}
const aSelection = {
  _id: q.literal('a')
  a: q.boolean()
}
const conditionB = {
  _id: q.literal('b')
  b: q.number()
}
const selectionTuple = [baseSelection, {
  '_id == "a"': aSelection
  '_id == "b"': bSelection
}]

type SomeEntity = TypeFromSelection<typeof baseSelection> 
| (TypeFromSelection<typeof baseSelection> & (
  | TypeFromSelection<typeof aSelection>
  | TypeFromSelection<typeof bSelection>
))

const query = groq('*')
  .filter('_type == "someEntity"')
  .grab(...selectionTuple)
```

This has its pitfalls tho:
- handling of standalone entities was a little sloppy (need to se a tuple that you spread into the `grab`)
- no support for default condition
- mixed support for empty base selection
- tough to infer the type from the `selectionTuple`
- doesn't handle types well where there's a default condition and empty base (ie sanity field is only ever an `a` entity or a `b` entity)